src/com/SocketSendQueue.cpp

```diff
@@ -0,0 +1,43 @@
+#include "SocketSendQueue.hpp"
+
+namespace precice
+{
+namespace com
+{
+namespace asio = boost::asio;
+
+void SendQueue::push(std::shared_ptr<Socket> sock, 
+        boost::asio::const_buffers_1 data, 
+        std::function<void()> callback)
+{
+  _itemQueue.push_back({sock, data, callback});
```

Comment:
I _think_ you can call `emplace_back` here and move the sock and data. Can you please check if that is possible?
```suggestion
  _itemQueue.emplace_back(std::move(sock), std::move(data), callback);
```

Reply:
Unfortunately this does not work. 
**TL;DR**: It should work but the standard library forbids it.
`SendItem` does not have any constructors other than the implicit ones. In the end `vector::emplace_back` tries to construct a new element by `::new (p) (sock, data, callback)`. This has to do with the way `std::allocator_traits::construct` works. I don't quite know why they don't use the uniform initialization syntax (since it was introduced with C++11 anyway), but my guess is that they want to avoid these strange cases where constructors with `std::initializer_list` take precedence over other constructors. Therefore no aggregate initialization is performed.

Reply:
In principle i could add the respective constructor to `SendItem` but copying a `shared_ptr` and a `const_buffers_1` is very cheap (the latter is just a `void*` with a `size_t`) anyway.

Reply:
That's a shame, but good to be aware of that! Thank you for the write-up.

You can move the shared ptr and the data member though :wink:

```suggestion
  _itemQueue.push_back({std::move(sock), std::move(data), callback});
```